File: Palindrome_Partitioning.cpp
#include <vector>
#include <string>
using std::vector;
using std::string;
using std::move;

class Solution {
public:
    vector<vector<string>> partition(const string &s) {
        vector<vector<string>> ans;
        if (s.empty()) {
            return move(ans);
        }

        vector<vector<bool>> is_pali;
        for (int i = 0; i < s.size(); ++i) {
            vector<bool> pali(s.size() - i + 1, false);
            is_pali.push_back(move(pali));
        }

        for (int i = 0; i < s.size(); ++i) {
            for (int beg = i, offset = 0; offset <= 1; ++offset) {
                int left = beg, right = offset;
                while (left >= 0 && right + left < s.size()) {
                    if (s[left] == s[right + left]) {
                        is_pali[left][right] = true;
                    } else {
                        break;
                    }
                    --left;
                    right += 2;
                }
            }
        }
        vector<string> stack;
        dfs(ans, is_pali, stack, s, 0);
        return move(ans);
    }

private:
    void dfs(vector<vector<string>> &ans, const vector<vector<bool>> &is_pali, vector<string> &stack,
             const string &s, int idx) {
        if (idx == s.size()) {
            ans.push_back(stack);
            return;
        }

        for (int i = idx; i < s.size(); ++i) {
            if (is_pali[idx][i - idx]) {
                stack.push_back(move(s.substr(idx, i - idx + 1)));
                dfs(ans, is_pali, stack, s, i + 1);
                stack.pop_back();
            }
        }
    }
};

File: Palindrome_Partitioning.py
class Solution:

    # @param s, a string
    # @return a list of lists of string
    def partition(self, s):
        if not s:
            return []
        ans = [[s[0]]]
        for c in s[1:]:
            tmp = [a + [c] for a in ans]
            news = []
            idx = len(ans[0]) - 1
            while idx >= 0:
                tmps = ''.join(ans[0][idx:]) + c
                if self.is_palindrome(tmps):
                    news.append(ans[:idx] + [tmps])
                idx -= 1
            ans = tmp + news
        return ans

    def is_palindrome(self, string):
        left, right = 0, len(string) - 1
        while left <= right:
            if string[left] != string[right]:
                return False
            left += 1
            right -= 1
        return True


if __name__ == '__main__':
    s = Solution()
    print s.partition('eegiicgaeadbcfacfhifdbiehbgejcaeggcgbahfcajfhjjdgj')
